sequencer/src/persistence/fs.rs

```diff
@@ -893,14 +934,322 @@ impl SequencerPersistence for Persistence {
         ))
     }
 
-    async fn migrate_consensus(
+    async fn append_da2(
+        &self,
+        proposal: &Proposal<SeqTypes, DaProposal2<SeqTypes>>,
+        _vid_commit: VidCommitment,
+    ) -> anyhow::Result<()> {
+        let mut inner = self.inner.write().await;
+        let view_number = proposal.data.view_number().u64();
+        let dir_path = inner.da2_dir_path();
+
+        fs::create_dir_all(dir_path.clone()).context("failed to create da dir")?;
+
+        let file_path = dir_path.join(view_number.to_string()).with_extension("txt");
+        inner.replace(
+            &file_path,
+            |_| {
+                // Don't overwrite an existing proposal, but warn about it as this is likely not
+                // intended behavior from HotShot.
+                tracing::warn!(view_number, "duplicate DA proposal");
+                Ok(false)
+            },
+            |mut file| {
+                let proposal_bytes = bincode::serialize(&proposal).context("serialize proposal")?;
+                file.write_all(&proposal_bytes)?;
+                Ok(())
+            },
+        )
+    }
+
+    async fn append_proposal2(
         &self,
-        _migrate_leaf: fn(Leaf) -> Leaf2,
-        _migrate_proposal: fn(
-            Proposal<SeqTypes, QuorumProposal<SeqTypes>>,
-        ) -> Proposal<SeqTypes, QuorumProposal2<SeqTypes>>,
+        proposal: &Proposal<SeqTypes, QuorumProposalWrapper<SeqTypes>>,
     ) -> anyhow::Result<()> {
-        // TODO: https://github.com/EspressoSystems/espresso-sequencer/issues/2357
+        self.append_quorum_proposal2(proposal).await
+    }
+
+    async fn migrate_anchor_leaf(&self) -> anyhow::Result<()> {
+        let mut inner = self.inner.write().await;
+
+        if inner.migrated.contains("anchor_leaf") {
+            tracing::info!("decided leaves already migrated");
+            return Ok(());
+        }
+
+        let decided_leaf2_path = inner.decided_leaf2_path();
+
+        fs::create_dir_all(decided_leaf2_path.clone())
+            .context("failed to create anchor leaf 2  dir")?;
+
+        let decided_leaf_path = inner.decided_leaf_path();
+        if !decided_leaf_path.is_dir() {
+            return Ok(());
+        }
+
+        tracing::warn!("migrating decided leaves..");
+        for entry in fs::read_dir(decided_leaf_path)? {
```

Comment:
Do I understand correctly that if the migration is interrupted it will just migrate everything again the next time? I don't think that's a problem, as long as things eventually get migrated and we don't end up in a state where the node can't start up because the migration fails.

Reply:
Yeah, we can keep the offsets after each batch upsert and resume from there too. I didn't think about that but can add it after doing some rewards work